### ISSUE TITLE
Routes with optional parameters seem to be matching substring routes with similar names

### ISSUE THREAD

**Issue by dantheother** [aurelia/router#575]
**I'm submitting a bug report**

* **Library Version:**
1.3.0 and whatever is in the gistrun

**Please tell us about your environment:**
* **Operating System:**
Windows 10

* **Node Version:**
610.1

* **NPM Version:**
5.6.0

* **Browser:**
all 

* **Language:**
all 

**Current behavior:**

Route foo/:id? and fooBar/:id?. Cannot navigate to /fooBar by itself, foo gets activated with an id parameter of Bar.

Gist is at https://gist.run/?id=7e5747845dbca68fec9ff717195a15cd - the gist only reproduces the programatic interaction with the router, but the same thing happens if you type /fooBar into the address bar as if the code asks the router for a route to /fooBar.

**Expected/desired behavior:**

* **What is the expected behavior?**
being able to have routes that are substrings of other routes

**Comment by davismj** [comment]
Excellent repro, thx @dantheother

**Comment by fkleuver** [comment]
I got [nerd sniped](https://xkcd.com/356/) by this..

When I was looking at the parsed states, there was an epsilon segment at the 'b' of 'foobar' which made it jump over the '/' segment and behave as if it were directly comparing a static segment to a dynamic one. AFAIK, the static one is meant to win in that case.

I submitted a [PR](https://github.com/aurelia/route-recognizer/pull/33) in `route-recognizer` that solves this particular case (with a unit test, and the others still pass as well) but probably not in a good way now that I think about it more.

I'm still leaving it there as food for thought. Maybe you have a different approach in mind, just wanted to share mine.
